### ISSUE TITLE
Deprecate `word` in `report()`

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#8050]
Should we deprecate `word`?

https://github.com/stylelint/stylelint/blob/31c24c4f017892f4348872625725a3db4661af77/lib/utils/report.mjs#L27

We are moving away from it in function of various changes around fixers, reported ranges, ...

The main issue with using `word` is that a single word can appear multiple times in a given bit of CSS text.

-------

I like the ease of use of the `word` API as it allows plugin authors to quickly create a rule that works well enough in most cases.

But I am unsure if we should have an API that is known to have such issues

**Comment by Mouvedia** [comment]
I recommend to start with reviewing our usage of `word`.
It is currently being used by 35 core rules.
e.g. check if we have cases where always picking the first match induces bugs

If we have bugs then it will warrant deprecating `word` in the next major.
i.e. reviewing our rules is way more valuable for our users than deprecating `word`

**Comment by romainmenke** [comment]
We should first decide if we want to consider deprecating `word` at all :)

For the core rules we are consistently moving away from using `word` and each time we do, this fixes bugs.

But despite that it might still be a valuable API to offer rule authors?
I find it hard to balance the easy of use of this API against the chance of encountering real issues because of it.

Core rules are widely used and should be held to a very high standard.
But some rule authors might not need such accuracy?

----

> check if we have cases where always picking the first match induces bugs

The only scenario where `word` might be acceptable is single word constructs:
- a name of an `at-rule` e.g. `media` for `@media screen {}`
- a property e.g. `color` for `.foo { color: red; }`

Any other case will result in bugs.

But these bugs are of a very specific nature.
- all report will be valid
- all report will point to a relevant source location
- fixing errors one by one will eventually reveal all errors at their correct source position

The only issue is that initially multiple errors will be reported at the first instance in source.

-------

So to summarize
- `word` is fairly handy when writing rules
- `word` can produce reports on incorrect source locations
- end users can easily work out where all issues really are by resolving issues one by one

this is about technical purity vs. ease of use

**Comment by ybiquitous** [comment]
@romainmenke Thanks for raising this up and for the good summary. 👍🏼 

I really understand your motivation. And we should avoid `word` in core rules as much as possible. But to be honest, deprecating `word` might not be accepted by light rule authors because using `word` makes it very easy to write a custom rule in private projects. `word` enables users to be free from index calculation.

FYI, PostCSS also supports `word`:
https://postcss.org/api/#warningoptions

**Comment by romainmenke** [comment]
> `word` enables users to be free from index calculation.

This in particular is what I like about this feature :)

---

Thank you for considering this.
Let's close this for now 🙇
